### PR TITLE
Use IProjectThreadingService in NuGetUIThreadHelper

### LINKED ISSUE #4185: Nuget.Client should use CPS JoinableTaskFactory when interacting with CPS to prevent deadlocks
Currently Nuget.Client uses the Shell `JoinableTaskFactory` when interacting with CPS projects [NuGetUIThreadHelper](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/PackageManagement.UI/Utility/NuGetUIThreadHelper.cs). This is causing deadlocks due to project lock issues. The CPS JTF has knowledge of the project lock and logic to prevent these issues. You can get the CPS version of JTF by importing the `IProjectThreadingService` from CPS. Let me know if you need assistance in the specifics of getting the `IProjectThreadingService`.

I am not sure how many deadlocks this is causing, but I have identified at least one when NuGet eventually sets an item property from within an `ExecuteAction`.

### PR DESCRIPTION
**Pull request by emgarten**
Update the NuGetUIThreadHelper to use IProjectThreadingService for VS 2017 and IThreadHandling for VS 2015. These JoinableTaskFactories are able to work with the CPS project system used by .NETCore projects and avoid hangs.

NuGetUIThreadHelper will now get a JTF from the ProjectService based on the version of VS.

Everywhere I could find that might access a CPS project has been moved to use NuGetUIThreadHelper, and in my testing this fixes the hang.

Other places in the code still use the shell JTF for UI only operations.

Fixes https://github.com/NuGet/Home/issues/4185

//cc @jviau @alpaix @zhili1208 @rrelyea @jainaashish @mishra14 @rohit21agrawal @AArnott

### CODE REVIEW

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPackageInstaller.cs [hunk 1]
```diff
@@ -20,6 +20,7 @@
 using NuGet.Versioning;
 using NuGet.VisualStudio.Implementation.Resources;
 using Task = System.Threading.Tasks.Task;
+using NuGet.PackageManagement.UI;
```
Comment (AArnott) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPackageInstaller.cs]: Ordering?

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPackageInstallerServices.cs [hunk 2]
```diff
@@ -18,6 +18,7 @@
 using NuGet.Protocol.Core.Types;
 using NuGet.Versioning;
 using NuGet.VisualStudio.Implementation.Resources;
+using NuGet.PackageManagement.UI;
```
Comment (AArnott) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPackageInstallerServices.cs]: Sorting again

**Review by AArnott — APPROVED** [review]
Looks reasonable.

File: src/NuGet.Clients/VisualStudio.Facade/NuGetUIThreadHelper.cs [hunk 3]
```diff
@@ -3,22 +3,54 @@
...
+        /// This overrides the default VsTaskLibraryHelper.ServiceInstance JTF.
+        /// </summary>
+        public static void SetJoinableTaskFactoryFromService(IServiceProvider serviceProvider)
+        {
+            var componentModel = serviceProvider.GetService(typeof(SComponentModel)) as IComponentModel;
+            var projectServiceAccessor = componentModel.GetService<IProjectServiceAccessor>();
+
+#if VS14
+            // Use IThreadHandling.AsyncPump for Visual Studio 2015
+            ProjectService projectService = projectServiceAccessor.GetProjectService();
```
Comment (alpaix) [inline src/NuGet.Clients/VisualStudio.Facade/NuGetUIThreadHelper.cs]: use `var`

File: src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs [hunk 4]
```diff
@@ -158,6 +158,9 @@ private IVsMonitorSelection VsMonitorSelection
             _dteEvents = _dte.Events.DTEEvents;
             _dteEvents.OnBeginShutdown += OnBeginShutDown;
 
+            // Set the JoinableTaskFactory for the current version of visual studio.
+            NuGetUIThreadHelper.SetJoinableTaskFactoryFromService(ServiceLocator.PackageServiceProvider);
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs]: don't use `ServiceLocator`. `PackageServiceProvider` is just `NuGetPackage` so simply pass `this`.
And also `componentModel` is fetched already, consider passing it instead.

**Review by alpaix — CHANGES_REQUESTED** [review]
`NuGetUIThreadHelper` should be initialized by the `RestoreManagerPackage` as well. Whichever package is loaded first.

And then consider lazy initialization of the helper as this change can bring perf down again.

File: src/NuGet.Clients/VisualStudio.Facade/NuGetUIThreadHelper.cs [hunk 5]
```diff
@@ -3,22 +3,54 @@
...
+            IThreadHandling threadHandling = projectService.Services.ThreadingPolicy;
+            _joinableTaskFactory = service.AsyncPump;
```
Comment (jviau) [inline src/NuGet.Clients/VisualStudio.Facade/NuGetUIThreadHelper.cs]: Did you mean to use `threadHandling` here and not `service` ?

File: src/NuGet.Clients/VisualStudio.Facade/NuGetUIThreadHelper.cs [hunk 6]
```diff
@@ -3,22 +3,54 @@
...
+#else
+            // Use IProjectService for Visual Studio 2017
+            IProjectService projectService = projectServiceAccessor.GetProjectService();
```
Comment (jainaashish) [inline src/NuGet.Clients/VisualStudio.Facade/NuGetUIThreadHelper.cs]: Couldn't we take `projectService` and `threadHandling `out of this codition and just use `AsyncPump `or `JoinableTaskFactory `based on VS version?

File: src/NuGet.Clients/VsConsole/PowerShellHost/NuGetConsole.Host.PowerShell.csproj [hunk 7]
```diff
@@ -105,6 +105,10 @@
       <Project>{306cddfa-ff0b-4299-930c-9ec6c9308160}</Project>
       <Name>NuGet.PackageManagement.VisualStudio</Name>
     </ProjectReference>
+    <ProjectReference Include="..\..\VisualStudio.Facade\NuGet.VisualStudio.Facade.csproj">
```
Comment (jainaashish) [inline src/NuGet.Clients/VsConsole/PowerShellHost/NuGetConsole.Host.PowerShell.csproj:108]: why do we need to add this ProjectReference? I don't see any other change in PowerShellHost project...

File: src/NuGet.Clients/VisualStudio.Facade/NuGetUIThreadHelper.cs (hunk 6 above)
Comment (emgarten) [reply src/NuGet.Clients/VisualStudio.Facade/NuGetUIThreadHelper.cs]: That API isn't currently in our dev14 build. Also, the threading docs say to use this for dev15.

File: src/NuGet.Clients/VsConsole/PowerShellHost/NuGetConsole.Host.PowerShell.csproj (hunk 7 above)
Comment (emgarten) [reply src/NuGet.Clients/VsConsole/PowerShellHost/NuGetConsole.Host.PowerShell.csproj:108]: ``NuGetUIThreadHelper`` moved from the UI (already referenced) to the Facade (one level down)

File: src/NuGet.Clients/VisualStudio.Facade/NuGetUIThreadHelper.cs (hunk 5 above)
Comment (emgarten) [reply src/NuGet.Clients/VisualStudio.Facade/NuGetUIThreadHelper.cs]: yes, good catch!

File: src/NuGet.Clients/VisualStudio.Facade/NuGetUIThreadHelper.cs (hunk 3 above)
Comment (emgarten) [reply src/NuGet.Clients/VisualStudio.Facade/NuGetUIThreadHelper.cs]: I'm going to leave this as is. With the compiler flags it is harder to determine these types and there is a key difference in that one is ``ProjectService`` and the other is ``IProjectService``

File: src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs (hunk 4 above)
Comment (emgarten) [reply src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs]: Good call, I've updated this to use the already created componentModel reference

**Review by jainaashish — APPROVED** [review]

**Review by alpaix — APPROVED** [review]

**Pull request closed without merging**